Regression: `NPE` on `null` extra property with 8.14 RC1

In Gradle 8.14 RC1, it is no longer possible to create extension properties with a `null` value:

```groovy
project.ext {
     aProperty = null as Integer
}
```

### Expected Behavior

The property is registered with the `ext` container (i.e. `ext.has("aProperty") == true && ext.properties.containsKey("aProperty")`).

### Actual Behavior

```
Caused by: java.lang.NullPointerException: null value in entry: aProperty=null
```

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/extensibility/DefaultExtraPropertiesExtension.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/extensibility/DefaultExtraPropertiesExtension.java
@@ -109,20 +109,21 @@ public class DefaultExtraPropertiesExtension extends GroovyObjectSupport impleme
 
     @Override
     public Map<String, Object> getProperties() {
+        // Must return a mutable map to preserve contract
         // TODO:configuration-cache use a tracking map here
         if (storage == null) {
-            return gradleProperties;
+            return new HashMap<>(gradleProperties);
         }
-        ImmutableMap.Builder<String, Object> builder = ImmutableMap.builderWithExpectedSize(storage.size() + gradleProperties.size());
-        builder.putAll(storage);
+        Map<String, Object> properties = new HashMap<>(storage.size() + gradleProperties.size());
+        properties.putAll(storage);
         for (Map.Entry<String, Object> entry : gradleProperties.entrySet()) {
             if (!storage.containsKey(entry.getKey())) {
                 // TODO:configuration-cache track Gradle property lookup
 //                onGradlePropertyLookup(entry.getKey());
-                builder.put(entry.getKey(), entry.getValue());
+                properties.put(entry.getKey(), entry.getValue());
             }
         }
-        return builder.build();
+        return properties;
     }
 
     @SuppressWarnings("rawtypes")

--- a/platforms/core-configuration/model-core/src/test/groovy/org/gradle/internal/extensibility/ExtraPropertiesExtensionTest.groovy
+++ b/platforms/core-configuration/model-core/src/test/groovy/org/gradle/internal/extensibility/ExtraPropertiesExtensionTest.groovy
@@ -19,6 +19,7 @@ package org.gradle.internal.extensibility
 import org.gradle.api.Project
 import org.gradle.api.plugins.ExtraPropertiesExtension
 import org.gradle.testfixtures.ProjectBuilder
+import spock.lang.Issue
 import spock.lang.Specification
 
 /**
@@ -188,6 +189,26 @@ abstract class ExtraPropertiesExtensionTest<T extends ExtraPropertiesExtension> 
 
         then:
         notThrown(Exception)
+    }
+
+    @Issue('https://github.com/gradle/gradle/issues/33074')
+    def "can add null properties using ext block notation"() {
+        given:
+        Project project = ProjectBuilder.builder().build()
+
+        when:
+        project.ext {
+            aProperty = null as Integer
+        }
+
+        then:
+        notThrown(Exception)
+
+        and:
+        project.ext.has 'aProperty'
+
+        and:
+        project.ext.properties.containsKey('aProperty')
     }
 
     def "can use [] notation to get and set"() {

